data_model/src/events/pipeline.rs

```diff
@@ -1,75 +1,65 @@
 //! Pipeline events.
 
 #[cfg(not(feature = "std"))]
-use alloc::{format, string::String, vec::Vec};
+use alloc::{boxed::Box, format, string::String, vec::Vec};
 
-use getset::Getters;
-use iroha_crypto::Hash;
+use iroha_crypto::HashOf;
 use iroha_data_model_derive::model;
 use iroha_macro::FromVariant;
 use iroha_schema::IntoSchema;
 use parity_scale_codec::{Decode, Encode};
 use serde::{Deserialize, Serialize};
-use strum::EnumDiscriminants;
 
 pub use self::model::*;
+use crate::transaction::SignedTransaction;
 
 #[model]
 pub mod model {
+    use getset::Getters;
+
     use super::*;
 
-    /// [`Event`] filter.
     #[derive(
         Debug,
         Clone,
-        Copy,
         PartialEq,
         Eq,
         PartialOrd,
         Ord,
-        Default,
-        Getters,
+        FromVariant,
         Decode,
         Encode,
-        Serialize,
         Deserialize,
+        Serialize,
         IntoSchema,
     )]
-    pub struct PipelineEventFilter {
-        /// If `Some::<EntityKind>`, filter by the [`EntityKind`]. If `None`, accept all the [`EntityKind`].
-        pub(super) entity_kind: Option<PipelineEntityKind>,
-        /// If `Some::<StatusKind>`, filter by the [`StatusKind`]. If `None`, accept all the [`StatusKind`].
-        pub(super) status_kind: Option<PipelineStatusKind>,
-        /// If `Some::<Hash>`, filter by the [`struct@Hash`]. If `None`, accept all the [`struct@Hash`].
-        // TODO: Can we make hash typed like HashOf<SignedTranasaction>?
-        pub(super) hash: Option<Hash>,
+    #[ffi_type(opaque)]
+    pub enum PipelineEventBox {
+        Transaction(TransactionEvent),
+        Block(BlockEvent),
     }
 
-    /// The kind of the pipeline entity.
     #[derive(
         Debug,
         Clone,
-        Copy,
         PartialEq,
         Eq,
         PartialOrd,
         Ord,
+        Getters,
         Decode,
         Encode,
         Deserialize,
         Serialize,
         IntoSchema,
     )]
     #[ffi_type]
-    #[repr(u8)]
-    pub enum PipelineEntityKind {
-        /// Block
-        Block,
-        /// Transaction
-        Transaction,
+    #[getset(get = "pub")]
+    pub struct BlockEvent {
```

Comment:
Hash of block might be helpful as well.

Reply:
can you tell me how? who can predict the hash of the block to filter by ahead of time?
I guess it's possible during block replay when restarting the blockchain but that seems of marginal utility